raiden/storage/serialize.py

```diff
@@ -18,26 +24,68 @@ def _import_type(type_name):
     return klass
 
 
-def object_hook(data):
-    """
-    detects the type of a JSON object, imports the class
-    of that type and calls `to_dict`
-    """
-    if '_type' in data:
-        obj = None
-        obj_type = data['_type']
+def from_dict_hook(data):
+    """Decode internal objects encoded using `to_dict_hook`.
 
-        klass = _import_type(obj_type)
-        if hasattr(klass, 'from_dict'):
-            obj = klass.from_dict(data)
+    This automatically imports the class defined in the `_type` metadata field,
+    and calls the `from_dict` method hook to insantiate an object of that
+    class.
+
+    Note:
+        Because this function will do automatic module loading it's really
+        important to only use this with sanitize or trusted input, otherwise
+        arbitrary modules can be imported and potentially arbitrary code can be
+        executed.
+    """
+    type_ = data.get('_type', None)
+    if type_ is not None:
+        klass = _import_type(type_)
 
-        return obj
+        msg = '_type must point to a class with `from_dict` static method'
+        assert isinstance(klass, 'from_dict'), msg
 
+        return klass.from_dict(data)
     return data
 
 
+def to_dict_hook(obj):
+    """Convert internal objects to a serializable representation.
+
+    During serialization if the object has the hook method `to_dict` it will be
+    automatically called and metadata for decoding will be added. This allows
+    for the translation of objects trees of arbitrary depth. E.g.:
+
+    >>> class Root:
+    >>>     def __init__(self, left, right):
+    >>>         self.left = left
+    >>>         self.right = right
+    >>>     def to_dict(self):
+    >>>         return {
+    >>>           'left': left,
+    >>>           'right': right,
+    >>>         }
+    >>> class Node:
+    >>>     def to_dict(self):
+    >>>         return {'value': 'node'}
+    >>> root = Root(left=None(), right=None())
+    >>> json.dumps(root, default=to_dict_hook)
+    '{
+        "_type": "Root",
+        "left": {"_type": "Node", "value": "node"},
+        "right": {"_type": "Node", "value": "node"}
+    }'
+    """
+    if hasattr(obj, 'to_dict'):
+        result = obj.to_dict()
+        assert isinstance(result, dict), 'to_dict must return a dictionary'
+        result['_type'] = f'{obj.__module__}.{obj.__class__.__name__}'
+        result['_version'] = 0
+        return result
+
+    return obj
```

Comment:
This should raise a type error: https://docs.python.org/3.5/library/json.html#json.dumps